File: zlt-business/saleops-optimizer/src/main/java/com/central/soo/model/entity/ModelVariable.java
package com.central.soo.model.entity;

import com.baomidou.mybatisplus.annotation.*;
import com.central.common.model.BaseEntity;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.math.BigDecimal;
import java.util.List;

/**
 * 模型变量实体
 * 对应数据库表：soo_model_variable
 * 
 * @author Portal 3.0 开发团队
 * @since 2024-12-19
 */
@Data
@EqualsAndHashCode(callSuper = false)
@TableName("soo_model_variable")
public class ModelVariable extends BaseEntity {

    /**
     * 主键ID
     */
    @TableId(value = "id", type = IdType.AUTO)
    private Long id;

    /**
     * 模型ID
     */
    @TableField("model_id")
    private Long modelId;

    /**
     * 变量编码
     */
    @TableField("variable_code")
    private String variableCode;

    /**
     * 变量名称
     */
    @TableField("variable_name")
    private String variableName;

    /**
     * 变量类型：INPUT/CALC/API
     */
    @TableField("variable_type")
    private String variableType;

    /**
     * 数据类型：NUMBER/DECIMAL/PERCENTAGE/CURRENCY/BOOLEAN/STRING
     */
    @TableField("data_type")
    private String dataType;

    /**
     * 单位
     */
    @TableField("unit")
    private String unit;

    /**
     * 父级变量ID，支持树形结构
     */
    @TableField("parent_id")
    private Long parentId;

    /**
     * 默认值
     */
    @TableField("default_value")
    private BigDecimal defaultValue;

    /**
     * 最小值
     */
    @TableField("min_value")
    private BigDecimal minValue;

    /**
     * 最大值
     */
    @TableField("max_value")
    private BigDecimal maxValue;

    /**
     * 计算公式
     */
    @TableField("calculation_formula")
    private String calculationFormula;

    /**
     * API配置
     */
    @TableField("api_config")
    private String apiConfig;

    /**
     * 约束条件公式
     */
    @TableField("constraint_formula")
    private String constraintFormula;

    /**
     * 显示顺序
     */
    @TableField("display_order")
    private Integer displayOrder;

    /**
     * 是否必填
     */
    @TableField("is_required")
    private Boolean isRequired;

    /**
     * 是否关键指标
     */
    @TableField("is_key_indicator")
    private Boolean isKeyIndicator;

    /**
     * 是否显示
     */
    @TableField("is_visible")
    private Boolean isVisible;

    /**
     * 验证规则
     */
    @TableField("validation_rules")
    private String validationRules;

    /**
     * 变量描述
     */
    @TableField("description")
    private String description;

    /**
     * 帮助说明
     */
    @TableField("help_text")
    private String helpText;

    // 非数据库字段 - 用于树形结构展示
    /**
     * 子变量列表
     */
    @TableField(exist = false)
    private List<ModelVariable> children;

    /**
     * 层级深度
     */
    @TableField(exist = false)
    private Integer level;

    /**
     * 是否为叶子节点
     */
    @TableField(exist = false)
    private Boolean isLeaf;

    /**
     * 父变量名称
     */
    @TableField(exist = false)
    private String parentName;

    /**
     * 子变量数量
     */
    @TableField(exist = false)
    private Integer childrenCount;

    // 变量类型常量
    public static final String TYPE_INPUT = "INPUT";
    public static final String TYPE_CALC = "CALC";
    public static final String TYPE_API = "API";
    public static final String TYPE_CALC_FACTORS = "CALC_FACTORS";

    // 数据类型常量
    public static final String DATA_TYPE_NUMBER = "NUMBER";
    public static final String DATA_TYPE_DECIMAL = "DECIMAL";
    public static final String DATA_TYPE_PERCENTAGE = "PERCENTAGE";
    public static final String DATA_TYPE_CURRENCY = "CURRENCY";
    public static final String DATA_TYPE_BOOLEAN = "BOOLEAN";
    public static final String DATA_TYPE_STRING = "STRING";

    /**
     * 判断是否为根节点
     */
    public boolean isRoot() {
        return parentId == null || parentId == 0;
    }

    /**
     * 判断是否为叶子节点
     */
    public boolean isLeafNode() {
        return children == null || children.isEmpty();
    }

    /**
     * 获取变量类型显示名称
     */
    public String getVariableTypeDisplayName() {
        switch (variableType) {
            case TYPE_INPUT:
                return "输入";
            case TYPE_CALC:
                return "计算";
            case TYPE_API:
                return "API";
            case TYPE_CALC_FACTORS:
                return "计算因子";
            default:
                return variableType;
        }
    }

    /**
     * 判断变量类型是否需要用户输入
     * INPUT和API类型不参与实例变量填写
     * CALC_FACTORS和CALC类型需要用户填写
     */
    public boolean isUserInputRequired() {
        return TYPE_CALC_FACTORS.equals(variableType) || TYPE_CALC.equals(variableType);
    }

    /**
     * 判断变量类型是否显示计算表达式
     * 只有CALC类型才显示计算表达式
     */
    public boolean shouldShowCalculationFormula() {
        return TYPE_CALC.equals(variableType);
    }

    /**
     * 判断变量类型是否显示约束表达式
     * 所有类型都显示约束表达式（如果存在）
     */
    public boolean shouldShowConstraintFormula() {
        return constraintFormula != null && !constraintFormula.trim().isEmpty();
    }

    /**
     * 获取数据类型显示名称
     */
    public String getDataTypeDisplayName() {
        switch (dataType) {
            case DATA_TYPE_NUMBER:
                return "数字";
            case DATA_TYPE_DECIMAL:
                return "小数";
            case DATA_TYPE_PERCENTAGE:
                return "百分比";
            case DATA_TYPE_CURRENCY:
                return "货币";
            case DATA_TYPE_BOOLEAN:
                return "布尔值";
            case DATA_TYPE_STRING:
                return "字符串";
            default:
                return dataType;
        }
    }
} 

File: zlt-web/portal-web/src/services/modelVariable.ts
import { request } from '@/utils/request';
import { API_ENDPOINTS } from '@/config/api';

const API_BASE = `${API_ENDPOINTS.SOO}/api/soo/v2/variables`;

export interface ModelVariable {
  id: number;
  modelId: number;
  variableName: string;
  variableCode: string;
  variableType: 'INPUT' | 'CALC' | 'API' | 'CALC_FACTORS';
  dataType: 'NUMBER' | 'DECIMAL' | 'PERCENTAGE' | 'CURRENCY' | 'BOOLEAN' | 'STRING';
  defaultValue?: number;
  minValue?: number;
  maxValue?: number;
  unit?: string;
  description?: string;
  calculationFormula?: string; // 匹配后端字段名
  apiConfig?: string; // 匹配后端字段名
  isRequired: boolean;
  isKeyIndicator?: boolean;
  isVisible: boolean;
  validationRules?: string;
  displayOrder: number;
  helpText?: string;
  parentId?: number; // 新增：父级变量ID，支持树形结构
  constraintFormula?: string; // 新增：约束条件公式
  children?: ModelVariable[]; // 新增：子变量列表，用于树形展示
  level?: number; // 新增：层级深度
  isLeaf?: boolean; // 新增：是否为叶子节点
  expanded?: boolean; // 新增：是否展开（前端状态）
  createdAt: string;
  updatedAt: string;
}

export interface VariableFormData {
  modelId: number;
  variableName: string;
  variableCode: string;
  variableType: 'INPUT' | 'CALC' | 'API' | 'CALC_FACTORS';
  dataType: 'NUMBER' | 'DECIMAL' | 'PERCENTAGE' | 'CURRENCY' | 'BOOLEAN' | 'STRING';
  defaultValue?: number;
  minValue?: number;
  maxValue?: number;
  unit?: string;
  description?: string;
  formulaExpression?: string; // 前端表单字段名，会映射到calculationFormula
  apiUrl?: string; // 前端表单字段名，会映射到apiConfig
  isRequired: boolean;
  isVisible: boolean;
  displayOrder: number;
  validationRules?: string;
  parentId?: number; // 新增：父级变量ID
  constraintFormula?: string; // 新增：约束条件公式
}

export interface PageParams {
  current?: number;
  pageSize?: number;
  modelId?: number;
  keyword?: string;
  variableType?: string;
  dataType?: string;
}

export interface PageResult<T> {
  data: T[];
  count: number;
  page: number;
  size: number;
  pages: number;
}

export interface ApiResponse<T = any> {
  resp_code: number;
  datas: T;
  resp_msg: string;
}

/**
 * 模型变量管理API服务
 */
export class ModelVariableAPI {
  
  /**
   * 获取模型变量分页列表
   */
  static async getVariables(params: PageParams): Promise<PageResult<ModelVariable>> {
    const response = await request<ApiResponse<PageResult<ModelVariable>>>(API_BASE, {
      method: 'GET',
      params: {
        page: params.current || 1,
        size: params.pageSize || 20,
        modelId: params.modelId,
        keyword: params.keyword,
        variableType: params.variableType,
        dataType: params.dataType,
      },
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '获取变量列表失败');
  }

  /**
   * 根据ID获取变量详情
   */
  static async getVariableById(id: number): Promise<ModelVariable> {
    const response = await request<ApiResponse<ModelVariable>>(`${API_BASE}/${id}`, {
      method: 'GET',
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '获取变量详情失败');
  }

  /**
   * 创建模型变量
   */
  static async createVariable(data: VariableFormData): Promise<ModelVariable> {
    const response = await request<ApiResponse<ModelVariable>>(API_BASE, {
      method: 'POST',
      data,
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '创建变量失败');
  }

  /**
   * 更新模型变量
   */
  static async updateVariable(id: number, data: VariableFormData): Promise<ModelVariable> {
    const response = await request<ApiResponse<ModelVariable>>(`${API_BASE}/${id}`, {
      method: 'PUT',
      data,
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '更新变量失败');
  }

  /**
   * 删除模型变量
   */
  static async deleteVariable(id: number): Promise<boolean> {
    const response = await request<ApiResponse<boolean>>(`${API_BASE}/${id}`, {
      method: 'DELETE',
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '删除变量失败');
  }

  /**
   * 批量删除变量
   */
  static async batchDeleteVariables(ids: number[]): Promise<boolean> {
    const response = await request<ApiResponse<boolean>>(`${API_BASE}/batch-delete`, {
      method: 'POST',
      data: { ids },
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '批量删除变量失败');
  }

  /**
   * 根据模型ID获取变量列表
   */
  static async getVariablesByModelId(modelId: number): Promise<ModelVariable[]> {
    const response = await request<ApiResponse<ModelVariable[]>>(`${API_BASE}/model/${modelId}`, {
      method: 'GET',
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '获取模型变量失败');
  }

  /**
   * 验证变量公式
   */
  static async validateFormula(modelId: number, formula: string): Promise<any> {
    const response = await request<ApiResponse<any>>(`${API_BASE}/validate-formula`, {
      method: 'POST',
      data: { modelId, formula },
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '验证公式失败');
  }

  /**
   * 复制变量
   */
  static async cloneVariable(
    id: number, 
    newVariableCode: string, 
    newVariableName: string
  ): Promise<ModelVariable> {
    const response = await request<ApiResponse<ModelVariable>>(`${API_BASE}/${id}/clone`, {
      method: 'POST',
      data: {
        newVariableCode,
        newVariableName,
      },
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '复制变量失败');
  }

  /**
   * 批量更新变量顺序
   */
  static async updateVariableOrder(variables: { id: number; displayOrder: number }[]): Promise<boolean> {
    const response = await request<ApiResponse<boolean>>(`${API_BASE}/update-order`, {
      method: 'PUT',
      data: { variables },
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '更新变量顺序失败');
  }

  /**
   * 获取变量统计信息
   */
  static async getVariableStatistics(modelId: number): Promise<any> {
    const response = await request<ApiResponse<any>>(`${API_BASE}/statistics/${modelId}`, {
      method: 'GET',
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '获取变量统计失败');
  }

  /**
   * 获取变量树形结构
   */
  static async getVariableTree(modelId: number): Promise<ModelVariable[]> {
    const response = await request<ApiResponse<ModelVariable[]>>(`${API_BASE}/tree/${modelId}`, {
      method: 'GET',
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '获取变量树失败');
  }

  /**
   * 验证约束条件
   */
  static async validateConstraint(
    modelId: number, 
    parentId: number, 
    constraintFormula: string
  ): Promise<any> {
    const response = await request<ApiResponse<any>>(`${API_BASE}/validate-constraint`, {
      method: 'POST',
      data: { modelId, parentId, constraintFormula },
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '验证约束条件失败');
  }

  /**
   * 获取变量的子变量列表
   */
  static async getChildVariables(parentId: number): Promise<ModelVariable[]> {
    const response = await request<ApiResponse<ModelVariable[]>>(`${API_BASE}/children/${parentId}`, {
      method: 'GET',
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '获取子变量失败');
  }

  /**
   * 移动变量到新的父级
   */
  static async moveVariable(id: number, newParentId: number | null): Promise<boolean> {
    const response = await request<ApiResponse<boolean>>(`${API_BASE}/${id}/move`, {
      method: 'PUT',
      data: { parentId: newParentId },
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '移动变量失败');
  }

  /**
   * 获取变量依赖关系
   */
  static async getVariableDependencies(variableId: number): Promise<any[]> {
    const response = await request<ApiResponse<any[]>>(`${API_BASE}/${variableId}/dependencies`, {
      method: 'GET',
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '获取变量依赖关系失败');
  }



  /**
   * 导出变量配置
   */
  static async exportVariables(modelId: number): Promise<string> {
    const response = await request<ApiResponse<string>>(`${API_BASE}/export/${modelId}`, {
      method: 'GET',
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '导出变量配置失败');
  }

  /**
   * 导入变量配置
   */
  static async importVariables(modelId: number, configJson: string): Promise<ModelVariable[]> {
    const response = await request<ApiResponse<ModelVariable[]>>(`${API_BASE}/import/${modelId}`, {
      method: 'POST',
      data: { configJson },
    });

    if (response.resp_code === 0) {
      return response.datas;
    }
    throw new Error(response.resp_msg || '导入变量配置失败');
  }
} 